File: dotnet/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/Warrior.cs
namespace DecoratorDesignPattern.Example1;

public class Warrior : ICharacter
{
    private readonly string _name;

    public Warrior(string name)
    {
        _name = name;
    }

    public virtual string GetName() => _name;
    public virtual int GetHealth() => 100;
    public virtual int GetAttackPower() => 20;
    public virtual int GetDefense() => 10;
    public virtual int GetSpeed() => 15;

    public virtual string GetDescription()
    {
        return $"{_name} (Savaşçı)";
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/src/Warrior.ts
import { ICharacter } from './ICharacter';

export class Warrior implements ICharacter {
  constructor(private readonly name: string) {}

  getName(): string { return this.name; }
  getHealth(): number { return 100; }
  getAttackPower(): number { return 20; }
  getDefense(): number { return 10; }
  getSpeed(): number { return 15; }
  getDescription(): string { return `${this.name} (Savaşçı)`; }
}
